### ISSUE TITLE
Proxy with caddy

### ISSUE THREAD

**Issue by robertgzr** [glowing-bear/glowing-bear#840]
I want to use [caddy](https://github.com/mholt/caddy) to proxy my weechat instance.
I've tried almost any configuration but I can't seem to get it to work...

```
irc.example.com {
  proxy /weechat localhost:9000 {
    header_upstream Host {host}
    header_upstream X-Real-IP {remote}
    header_upstream X-Forwarded-For {remote}
    header_upstream X-Forwarded-Proto {scheme}
    header_upstream Connection {>Connection}
    header_upstream Upgrade {>Upgrade}
  }
}
```

I get the following error messages in the developer tools console:

```
min.js:2 | WebSocket connection to 'wss://irc.example.com/weechat' failed: Error during WebSocket handshake: Unexpected response code: 400
```

```
angular.min.js:107 | Relay error Event {isTrusted: true, type: "error", target: WebSocket, currentTarget: WebSocket, eventPhase: 2…}
```

Dos anybody have any ideas what I am doing wrong or what to try next?

**Comment by lorenzhs** [comment]
Sorry, I don't use caddy. Probably better to ask in a caddy-related IRC channel? From a quick google search this was fixed in https://github.com/mholt/caddy/pull/987 so you probably need version 0.9.1 or newer of caddy.

Does your weechat get a connection (output in the core buffer)? What does `/relay raw` show?

Oh and please add it to https://github.com/glowing-bear/glowing-bear/wiki/Proxying-WeeChat-relay-with-a-web-server once you've figured it out :)

**Comment by robertgzr** [comment]
I was thinking that the you people here might have an idea what the underlying reason is from the console warnings...
But I will ask in the caddy-related channels too.

This is the `/relay raw` output:
```
  | User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.143 Safari/537.36
  | Accept-Encoding: gzip, deflate, sdch, br
  | Accept-Language: en-US,en;q=0.8,de;q=0.6
  | Cache-Control: no-cache
  | Connection: Upgrade
  | Dnt: 1
  | Origin: https://www.glowing-bear.org
  | Pragma: no-cache
  | Sec-Websocket-Extensions: permessage-deflate; client_max_window_bits
  | Sec-Websocket-Key: FAk5MmlhUzdPxkrl3WWacA==
  | Sec-Websocket-Version: 13
  | Upgrade: websocket
  | X-Forwarded-For: (my public ip)
  | X-Forwarded-Proto: https
  | X-Real-Ip: (my public ip)
  |
  <-- [1] weechat | HTTP/1.1 400 Bad Request
```

**Comment by lorenzhs** [comment]
That looks mostly correct. I suspect there's a mismatch in TLS usage in the forwarding. Is your relay encrypted (ssl.weechat) or not (weechat)? Does the host specification you're forwarding to reflect that?

**Comment by robertgzr** [comment]
Unencrypted relay.

The way I understand it weechat itself shouldn't expect any TLS at all right?
And caddy uses a letsencrypt certificate...

The connection settings for glowingbear are:
host: `irc.example.com` port: `443`

**Comment by lorenzhs** [comment]
Whether weechat expects TLS depends on how you configure it. That's why I listed `ssl.weechat` and `weechat` as the relay types above. If you configured it correctly then I guess you'll have to ask the caddy people why it fails. Where your weechat replies 400 bad request, mine says:
```
13:29 <-- [1] weechat │ HTTP/1.1 101 Switching Protocols
13:29                 │ Upgrade: websocket
13:29                 │ Connection: Upgrade
13:29                 │ Sec-WebSocket-Accept: qqKM3uRnEBoKZSoMzKe8RPkMf04=
```
and following that GB sends "init compression=zlib,password=hunter2" and "info version", etc etc.

**Comment by robertgzr** [comment]
Do you have any idea what the `400 Bad Request` might be about?

**Comment by lorenzhs** [comment]
@robertgzr did you get it to work?

**Comment by robertgzr** [comment]
@lorenzhs I had to setup a new machine anyway so I'm going to give this a try again on a fresh install.

I will report back when I'm done.

**Comment by robertgzr** [comment]
The problem was caused by a bug in weechat that should be fixed in version 1.7

weechat/weechat#888
mholt/caddy#1332

**Comment by lorenzhs** [comment]
Thanks for the follow-up! That was quite interesting
